### PR TITLE
Production triggers

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes:
Move block production triggering out of BlockProducer. 
Moves block consumption (suggest to block tree by default) out of BlockProducer.
Allows integration with MEV plugin.

## Types of changes

What types of changes does your code introduce?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

Need manual testing too (running validators) and POSDAO tests

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Api/Extensions/IConsensusPlugin.cs [hunk 1]
```diff
@@ -24,8 +25,10 @@ namespace Nethermind.Api.Extensions
 {
     public interface IConsensusPlugin : INethermindPlugin
     {
-        Task<IBlockProducer> InitBlockProducer(ITxSource? txSource = null);
+        Task<IBlockProducer> InitBlockProducer(IBlockProductionTrigger? blockProductionTrigger = null, ITxSource? additionalTxSource = null);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Api/Extensions/IConsensusPlugin.cs:28]: Ad

File: src/Nethermind/Nethermind.Blockchain/Producers/BlockProducer.cs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            return IsRunning() && (maxProducingInterval == null || _lastProducedBlockDateTime.AddSeconds(maxProducingInterval.Value) > DateTime.UtcNow);
+        }
+
+        private async Task<Block?> TryProduceAndAnnounceNewBlock(CancellationToken token, BlockHeader? parentHeader = null)
+        {
+            using CancellationTokenSource tokenSource = CancellationTokenSource.CreateLinkedTokenSource(token, _producerCancellationToken!.Token);
+            token = tokenSource.Token;
+            
+            Block? block = null;
+            if (await _producingBlockLock.WaitOneAsync(1000, token))
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Producers/BlockProducer.cs]: Can we add a comment to what happens when it is false? Maybe drbug diagnostic then?

File: src/Nethermind/Nethermind.Blockchain/Producers/BlockProducer.cs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            {
+                try
+                {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Producers/BlockProducer.cs]: What does it mean here that the blovk was not produced? Can we add a comment.

File: src/Nethermind/Nethermind.Clique.Test/CliqueBlockProducerTests.cs [hunk 4]
```diff
@@ -195,6 +193,8 @@ public On CreateNode(PrivateKey privateKey, bool withGenesisAlreadyProcessed = f
                     MainnetSpecProvider.Instance, 
                     _cliqueConfig,
                     nodeLogManager);
+
+                var suggester = new ProducedBlockSuggester(blockTree, blockProducer);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Clique.Test/CliqueBlockProducerTests.cs:197]: Not used?

File: src/Nethermind/Nethermind.Consensus.Clique/CliquePlugin.cs [hunk 5]
```diff
@@ -192,6 +192,9 @@ public Task InitRpcModules()
         }
 
         public string SealEngineType => Nethermind.Core.SealEngineType.Clique;
+        
+        //TODO: redo clique producer
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus.Clique/CliquePlugin.cs:196]: ToDo?

File: src/Nethermind/Nethermind.Blockchain/Producers/CompositeBlockProductionTrigger.cs [hunk 6]
```diff
@@ -16,34 +16,43 @@
...
         {
-            TriggerBlockProduction?.Invoke(sender, e);
+            _triggers.Add(trigger);
+            HookTrigger(trigger);
         }
+        
+        private void HookTrigger(IBlockProductionTrigger trigger) =>
+            trigger.TriggerBlockProduction += BlockProductionTriggerOnTriggerBlockProduction;
+
+        private void BlockProductionTriggerOnTriggerBlockProduction(object? sender, BlockProductionEventArgs e) =>
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Producers/CompositeBlockProductionTrigger.cs:46]: The name is ok? A lengthy palindrome

File: src/Nethermind/Nethermind.Blockchain/Producers/MultipleBlockProducer.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+namespace Nethermind.Blockchain.Producers
+{
+    public abstract class MultipleBlockProducer<T> : IBlockProducer where T : IBlockProducerInfo
+    {
+        private readonly IBlockProductionTrigger _blockProductionTrigger;
+        private readonly IBestBlockPicker _bestBlockPicker;
+        private readonly T[] _blockProducers;
+
+        protected MultipleBlockProducer(
+            IBlockProductionTrigger blockProductionTrigger,
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Producers/MultipleBlockProducer.cs:36]: Move to MEV plugin unless used in kther places

**Review by MarekM25 — COMMENTED** [review]
I proposed cosmetic changes. I wonder if we can test it on one of our producers before the merge... I run posdao tests on it. If tests pass, I will approve it.

**Review by tkstanczak — APPROVED** [review]
😍

File: src/Nethermind/Nethermind.Blockchain/Producers/MultipleBlockProducer.cs (hunk 7 above)
Comment (kristofgazso) [reply src/Nethermind/Nethermind.Blockchain/Producers/MultipleBlockProducer.cs:36]: i wouldnt, im going to have to use it in Account Abstraction for example

File: src/Nethermind/Nethermind.Api/Extensions/IConsensusPlugin.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Api/Extensions/IConsensusPlugin.cs:36]: It is dealt inside via composite pattern.

So we want the consensus to be in charge of how it adds transactiosn to blocks, as there are cases where there are consensus transactions that need to go first into the block (AuRa, POSDAO), but we also want to be able to inject some tx's from the outside (MEV), so I was debating 2 options:
1. additionalTxSource (named like that in order to not confuse that its a main tx source)
2. event to inject additional transactions

went with 1.

File: src/Nethermind/Nethermind.AuRa.Test/AuRaBlockProducerTests.cs [hunk 8]
```diff
@@ -99,20 +100,30 @@ private void InitProducer()
...
-                            BlockProcessingQueue,
                             Timestamper,
                             AuRaStepCalculator,
                             NullReportingValidator.Instance,
                             auraConfig,
                             new FollowOtherMiners(MainnetSpecProvider.Instance),
                             MainnetSpecProvider.Instance,
                             LimboLogs.Instance);
+
+                        var suggester = new ProducedBlockSuggester(BlockTree, AuRaBlockProducer);
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AuRa.Test/AuRaBlockProducerTests.cs:126]: ProducedBlockSuggester hooks to events, I can remove the variable which is not used, but prefer to have it there.

File: src/Nethermind/Nethermind.Blockchain.Test/Producers/DevBlockproducerTests.cs [hunk 9]
```diff
@@ -106,14 +106,14 @@ public void Test()
...
                 blockTree,
-                blockchainProcessor,
                 trigger,
                 timestamper,
                 specProvider,
                 new MiningConfig {Enabled = true},
                 LimboLogs.Instance);
 
             blockchainProcessor.Start();
+            var suggester = new ProducedBlockSuggester(blockTree, devBlockProducer);
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Blockchain.Test/Producers/DevBlockproducerTests.cs:116]: ProducedBlockSuggester hooks to events, I can remove the variable which is not used, but prefer to have it there.

File: src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs [hunk 10]
```diff
@@ -46,164 +49,216 @@ public abstract class BlockProducerBase : IBlockProducer
...
-            if (Logger.IsTrace)
-                Logger.Trace(
-                    $"Checking IsProducingBlocks: maxProducingInterval {maxProducingInterval}, _lastProducedBlock {_lastProducedBlockDateTime}, IsRunning() {IsRunning()}");
-            if (IsRunning() == false)
-                return false;
-            if (maxProducingInterval != null)
-                return _lastProducedBlockDateTime.AddSeconds(maxProducingInterval.Value) > DateTime.UtcNow;
-            else
-                return true;
+            if (Logger.IsTrace) Logger.Trace($"Checking IsProducingBlocks: maxProducingInterval {maxProducingInterval}, _lastProducedBlock {_lastProducedBlockDateTime}, IsRunning() {IsRunning()}");
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs:121]: Yes, thats why we have this heuristics based on intervals to see if blocks are actually produced.

File: src/Nethermind/Nethermind.Clique.Test/CliqueBlockProducerTests.cs (hunk 4 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Clique.Test/CliqueBlockProducerTests.cs:197]: ProducedBlockSuggester hooks to events, I can remove the variable which is not used, but prefer to have it there.

File: src/Nethermind/Nethermind.Consensus.Clique/CliquePlugin.cs (hunk 5 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus.Clique/CliquePlugin.cs:196]: Yes, wanted to indicate this will not work with Clique properly yet

**Review by kristofgazso — APPROVED** [review]

**Pull request merged**